### PR TITLE
Implement logcdf method for discrete distributions

### LINKED ISSUE #4331: Suggestion: CDF methods for discrete variables
Is there a reason why all discrete variables lack the logcdf method? 

Stan, for example, provides this functionality for several distributions (Poisson, Binomial, Negative binomial, Bernoulli)

Is this something people might be interested?

Best,
Ricardo

### PR DESCRIPTION
**Pull request by ricardoV94**
This PR adds `logcdf` methods to all univariate discrete distributions. Formulas were based in either scipy or wikipedia entries. For some distributions, I could not find specialized formulas and had to rely on summing up all the logps in `tt.arange(0, value+1)`.

Closes #4331

Unittests:
1. Using `check_logcdf` in all distributions with scipy counterpart.
2. Added a new type of unit test `check_selfconsistency_discrete_logcdf`, which tests that the logcdf is equivalent to adding all the logps starting from ~~zero~~ `domain.lower`. This provides coverage to the distributions that are not in scipy (not covered by 1.). I added it to all distributions, but maybe this is too redundant for those that are covered by 1. 

Some remaining issues / questions:
1. Should all `bound` calls in CDF methdos be replaced by `tt.switch`? Most `logcdf` methods for continous distributions seem to use `tt.switch`, except for the `Gamma` and `InverseGamma` distributions. I used `bound` whenever there were at least two conditions that needed checking leading no `-np.inf`, otherwise went with `tt.switch`. I am also not completely sure if this interferes with the recent #4377 PR.
2. Several `logcdf` methods do not work with multiple values (or even with an array containing a single value such as `np.array([1])`). In these cases, I excluded information in the docstrings about the possibility of computing the `logcdf` for multiple values, but this seems like a suboptimal solution. We should either fix it or raise a ValueError to explain the limitation to the user, as this seems to go against the expected behavior in most distributions. This occurs for two reasons:
   1. Computing logcdf by summing over the individual logps with `logsumexp(tt.arange(0, value+1), keepdims=False)` for the `BetaBinomial` and `HyperGeometric`. Any alternatives?
   2. Using the `incomplete_beta` function in the `Binomial` and `NegativeBinomial` (and `ZeroInflated` counterparts). This is also an issue in the current implementation of the `Beta` and `StudentT` distributions (see #4342). In addition, the unit tests for these functions seem to be very slow compared to those of other distributions. Is the `incomplete_beta` particularly slow?
3. Slightly different, the `logcdf` method of the `Poisson` distribution (and its `ZeroInflated` counterpart) fails with a C-assertion (exiting the python process altogether) when asked to evaluate multiple invalid values. This is also a problem in the `InverseGamma` distribution (see #4340). It will be solved once this Theano issue is fixed (see https://github.com/pymc-devs/Theano-PyMC/issues/224). **Update: I have found a temporary hack to "hide" this problem (which can be removed once those other issues are solved).**
4. Unittests: `check_logcdf` tests only for values in the output domain, ignoring the edge values, but we might want to test also for values at the edges as well as below or beyond the domain (which can be evaluated to either -np.inf or 0). For example `pm.Bernoulli(p=.1).logcdf([-1, 2]) -> [-np.inf, 0]`. Should we use more comprehensive extra-domain checks? **Update: This is now implemented in #4393.**

Other changes not directly related to this PR:
1. Added a missing `pymc3_matches_scipy` test to the `BetaBinomial`. Was there a reason why this was missing? 
2. Changed the order of the `logp` and `random` methods in the `DiscreteWeibull` distribution to be in line with the rest of the library.
3. Removed unused local variables in the init methods of `DiscreteWeibull` and `ZeroInflatedPoisson` distributions.

### CODE REVIEW

**Comment by ricardoV94** [comment]
I am sorry for all the changes after opening the PR. I think it is now ready for review.

File: pymc3/distributions/discrete.py
Comment (AlexAndorra) [inline pymc3/distributions/discrete.py:466]: Should probably add the types for multiple values then. Something like the following?
```suggestion
        value: numeric or np.ndarray or theano.tensor
```

File: pymc3/distributions/discrete.py
Comment (AlexAndorra) [inline pymc3/distributions/discrete.py:580]:
```suggestion
        value: numeric or np.ndarray or theano.tensor
```

File: pymc3/distributions/discrete.py
Comment (AlexAndorra) [inline pymc3/distributions/discrete.py:605]:
```suggestion
        value: numeric or np.ndarray or theano.tensor
```

File: pymc3/distributions/discrete.py
Comment (AlexAndorra) [inline pymc3/distributions/discrete.py:720]:
```suggestion
        value: numeric or np.ndarray or theano.tensor
```

File: pymc3/distributions/discrete.py
Comment (AlexAndorra) [inline pymc3/distributions/discrete.py:1020]:
```suggestion
        value: numeric or np.ndarray or theano.tensor
```

File: pymc3/distributions/discrete.py
Comment (AlexAndorra) [inline pymc3/distributions/discrete.py:1291]:
```suggestion
        value: numeric or np.ndarray or theano.tensor
```

File: pymc3/distributions/discrete.py
Comment (AlexAndorra) [inline pymc3/distributions/discrete.py:1604]:
```suggestion
        value: numeric or np.ndarray or theano.tensor
```

**Review by AlexAndorra — CHANGES_REQUESTED** [review]
This is great, thanks @ricardoV94 !
This is a pretty big PR, so a second review by another core dev would be beneficial.

I agree that the `logcdf` methods that do not work with multiple values should raise a `ValueError` to explain the limitation to the user. Related to this, I added some comments to add the mutiple values' types (numpy array or theano tensor) to the docstrings

**Comment by ricardoV94** [comment]
Thanks for the review @AlexAndorra. I agree about the ValueError, although I have to check how to check that with Theano variables.

I was hoping someone would suggest a fix for those that use `tt.arange`. Scipy actually does a loop to return results for multiple values.

About the type hints in the docstring. What if we do a separate PR just for that, so that all logcdfs (discrete and continuous) are homogeneous? Right now, none of the other docstrings have it and they should as well.

File: pymc3/distributions/discrete.py
Comment (twiecki) [inline pymc3/distributions/discrete.py:470]: Needs a new-line.

File: pymc3/distributions/discrete.py
Comment (twiecki) [inline pymc3/distributions/discrete.py:464]: Needs a new-line.

Comment (twiecki) [reply pymc3/distributions/discrete.py:464]: Above Parameters.

**Review by twiecki — CHANGES_REQUESTED** [review]
A lot of the doc-strings miss new-lines before Parameters and Returns, otherwise this looks great to me!

**Comment by ricardoV94** [comment]
> A lot of the doc-strings miss new-lines before Parameters and Returns, otherwise this looks great to me!

Thanks for reviewing it. I will fix those.

Unrelated to that, does anyone know a good way to check that `value` is not a scalar in order to raise an informative ValueError in those methods that fail with arrays/tensors?

**Comment by twiecki** [comment]
You can try whether `np.isscalar` works on numpy and theano arrays.

**Comment by ricardoV94** [comment]
New commit fixes missing newlines in docstrings and raises informative `TypeError` in logcdf methods that only accept scalars. 

I would address @AlexAndorra suggestion of including more informative docstring type hints, as well as expanding `check_logcdf` unittest to make sure that logcdf methods that fail with nonscalar values raise a `TypeError` in #4393. The reason for this is that the logcdf methods of some continuous distributions also need to be changed and I think it would be easier to keep track / review those in that separate PR. Do you agree?

**Review by twiecki — APPROVED** [review]

**Comment by twiecki** [comment]
This is a great contribution, thanks @ricardoV94!

**Pull request merged**
